### PR TITLE
[CI] Build and test using Docker on Travis

### PR DESCRIPTION
**Pull request by achilleas-k**
Opening as draft for feedback, ideas, and help.

- I merged @gicmo's commits from #771 for the setup and env vars.
- Moved the dockerfile to a new directory called `dockerfiles/`, which now contains two files: One for Ubuntu 18.04 and one for ArchLinux.
- Travis config specifies Dockerfile name.

Issues:
- Executable names in Ubuntu container differ from the Arch ones.  For instance, the versioned binary names like `g++-8` don't exist in Arch.  I don't want to have even longer chains of conditionals than we currently do.  Thoughts?

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -58,19 +45,47 @@ cache:
...
     then
+       export CC=gcc-8;
+       export CXX=g++-8;
+       export GCOV=gcov-8
        export CTEST_ARGS=-V;
     else
+       export CC=clang-7;
+       export CXX=clang++-7;
+       export GCOV=gcov-8;
        export CTEST_ARGS=-VV;
```
Comment (gicmo) [inline .travis.yml]: I guess lets just get rid of the explicit suffix and thus eliminate the branches as well

Comment (achilleas-k) [reply .travis.yml]: On it.

Comment (achilleas-k) [reply .travis.yml]: How do we handle the Ubuntu names?

Comment (gicmo) [reply .travis.yml]: We could use `update-alternatives` inside the container:

```
sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-7 1000
sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-7 1000
...
```

**Comment by achilleas-k** [comment]
We got green!

**Comment by achilleas-k** [comment]
Rebase was getting messy because I had merged @gicmo's older branch, so I merged master instead.  Marking as ready for review.

File: .travis.yml [hunk 2]
```diff
@@ -58,19 +45,41 @@ cache:
...
+    then
+      ./contrib/make-env.sh /tmp/docker.env;
+      docker build -f ./dockerfiles/${DOCKERFILE}.Dockerfile . -t local/nixio;
+      docker run --env-file /tmp/docker.env -d -i -t -v $(pwd):/src --name nixbuilder local/nixio;
+      docker exec nixbuilder ${CXX} --version;
+    elif [ "$TRAVIS_OS_NAME" == "osx" ];
+    then
+      brew update;
+      brew install cmake boost cppunit hdf5 yaml-cpp;
+      brew --env;
```
Comment (gicmo) [inline .travis.yml]: Isn't that done already by the `addons: homebrew` section above?

File: dockerfiles/ub1804.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+RUN update-alternatives --install /usr/bin/clang clang /usr/bin/clang-7 1000
+RUN update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-7 1000
+RUN update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-8 1000
+RUN update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-8 1000
+RUN update-alternatives --install /usr/bin/gcov gcov /usr/bin/gcov-8 1000
+
+RUN pip install cpp-coveralls && \
+    rm -rf /tmp/pip && \
+    rm -rf /root/.cache
```
Comment (gicmo) [inline dockerfiles/ub1804.Dockerfile]: We don't actually use that anymore, could be removed.

**Comment by achilleas-k** [comment]
Probably leftovers from merging.  I didn't really pay attention to the macos and coverage parts.

**Pull request merged**